rules/legacy/src/Rector/Include_/AddTopIncludeRector.php

```diff
@@ -0,0 +1,201 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\Legacy\Rector\Include_;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\Include_;
+use PhpParser\Node\Scalar\String_;
+use PhpParser\Node\Stmt\Class_;
+use PhpParser\Node\Stmt\Declare_;
+use PhpParser\Node\Stmt\DeclareDeclare;
+use PhpParser\Node\Stmt\GroupUse;
+use PhpParser\Node\Stmt\InlineHTML;
+use PhpParser\Node\Stmt\Namespace_;
+use PhpParser\Node\Stmt\Nop;
+use PhpParser\Node\Stmt\Use_;
+use PhpParser\Node\Stmt\UseUse;
+use Rector\Core\Exception\Rector\InvalidRectorConfigurationException;
+use Rector\Core\Rector\AbstractRector;
+use Rector\Core\RectorDefinition\CodeSample;
+use Rector\Core\RectorDefinition\RectorDefinition;
+use Rector\NodeTypeResolver\Node\AttributeKey;
+use ReflectionClass;
+
+/**
+ * @see https://github.com/rectorphp/rector/issues/3679
+ *
+ * @see \Rector\Legacy\Tests\Rector\Include_\AddTopIncludeRector\AddTopIncludeRectorTest
+ */
+final class AddTopIncludeRector extends AbstractRector
+{
+    /**
+     * @var bool
+     */
+    private $completed = false;
+
+    /**
+     * @var string
+     */
+    private $file = '';
+
+    /**
+     * @var string
+     */
+    private $lastFileProcessed = '';
+
+    /**
+     * @var int
+     */
+    private $type = 0;
+
+    /**
+     * @var string[]
+     */
+    private $classNodes = [
+        Declare_::class => true,
+        DeclareDeclare::class => true,
+        GroupUse::class => true,
+        InlineHTML::class => true,
+        Namespace_::class => true,
+        Use_::class => true,
+        UseUse::class => true,
+    ];
+
+    /**
+     * @var string[] to match
+     */
+    private $patterns = [];
+
+    /**
+     * @var array
+     */
+    private $configuration = [];
+
+    public function __construct(array $configuration = [])
+    {
+        $this->configuration = $configuration;
+    }
+
+    public function getDefinition(): RectorDefinition
+    {
+        return new RectorDefinition('Adds an include file at the top of matching files but not class definitions', [
+            new CodeSample(
+                '<?php
+
+if (isset($_POST["csrf"])) {
+    processPost($_POST);
+}',
+                '<?php
+
+include "autoloader.php";
+
+if (isset($_POST["csrf"])) {
+    processPost($_POST);
+}',
+                [
+                    '$configuration' => [
+                        'type' => 'TYPE_INCLUDE_FILE',
+                        'file' => 'autoload.php',
+                        'match' => ['filePathWildCards'],
+                    ],
+                ]
+            ),
+        ]);
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        $this->initialize();
+
+        return [Node::class];
```

Comment:
If you want to use somethign like `Node`, that's a sing this should be a `AbstractFileSystemRector`

Reply:
So finally had some time to look at this.  If I were to make it a AbstractFileSystemRector, then I would have to use BetterNodeFinder.  I can find a class node with that, but I also want to find the first executable statement and insert the include before that.  I currently just filter out nodes that are non-executable, then insert when I find a node that is not in that list. Not sure how I would do that using BetterNodeFinder.  There are some filters available, but not sure it is any better than what I am doing now.

Reply:
There is no need to use `BetterNodeFinder`, if you want to update top-level statements.

`foreach` over of nodes should do.

Could you update test fixtures with ~2 more samples that will cover edge-cases you discovered? 

Also 1 fixture with `skip_xxx.php.inc`, that checks the include is already there and to prevent double adding it.


<br>

I'll prepare short code sample for `AbstractFileSystemRector`, so it's easier to understand

Reply:
Here is a good example:

https://github.com/rectorphp/rector/blob/965cef98e218163cb43049c1e5ef0b32fa635eef/rules/legacy/src/Rector/FileSystem/FunctionToStaticMethodRector.php#L70-L73



```php
    public function refactor(SmartFileInfo $smartFileInfo): void
    {
        $nodes = $this->parseFileInfoToNodes($smartFileInfo);
        $fileStmts = $this->getFileOrNamespaceStmts($nodes);

        // 0. has already include? → skip
        foreach ($fileStmts as $fileStmt) {
           if ($this->isTopFileInclude(fileStmt)) {
               return;
           }
		}
      
        // 1. add include to right location
        foreach ($fileStmts as $fileStmt) {
           if (! $this->isLocationToAddInclude(fileStmt)) {
               continue;
           }
             
           // add it
           $fileStmts = ...;
		}

        // 2. print stmts back
        // ... mimic the process from the rule
   }
```

To add a stmt betweent other statements, you can use `array_splice`, but that was too magic for me and very hard to read/debug. So I use simple foreach:

```php
    $newFileStmts = [];
    foreach ($fileStmts as $fileStmt) {
        if ($this->isLocationToAddInclude(fileStmt)) {
           // add it
           $newFileStmts[] = new Expression(new Include_(...));               continue;
        }

        $newFileStmts[] = $fileStmt;
    }


// $this->print...( $newFileStmts)
```

Reply:
OK, thanks for the example!  They really help to understand the code.

Reply:
If that's too much to absorb now and you'll feel stressed, just add the test fixture and I'll merge it and do the change myself for this 1st case.

Reply:
Let me take another whack at it.  I think I understand what needs to be done.  The biggest issue for me is the PHPStan checks don't seem to run well under Windows 10. They also seem to reconfigure the method single line into multiple lines, which fail on github.  I prefer the single line format myself, so I was confused when it multi lined them.

I was also wondering about the string expression that could be in a include statement.  How would I specify something like __DIR__ . '/../file.php' in the config and get it converted into nodes?

Reply:
Hmm, I could probably just send the string through PhpParser, then add that expression to the include node.

Reply:
@phpfui 

> Hmm, I could probably just send the string through PhpParser, then add that expression to the include node.

Exactly! :+1:

Reply:
> The biggest issue for me is the PHPStan checks don't seem to run well under Windows 10.

I'm sorry to hear that. I actually don't know many CLI tools that work on Windows. It's often quite risky. I used Windows for PHP, but I do use Ubuntu for last 10 years and it's much easier as it just works. You might find more troubles with Rector on Windows in the future, just so you're aware of it.